File: Graph/GraphPractice.py
# Practice 10. Graph Representation 
from collections import deque
from pprint import pprint
CONSTRUCT = 'C'
IS_ADJACENT = 'I'
GET_NEIGHBORS = 'N'
BFS = 'B'
DFS = 'D'
REACHABILITY = 'R'
TOPOLOGICAL_SORT = 'T'
SHORTEST_PATH = 'S'

class Graph:
  # TODO. Define a constructor and proper methods
  def __init__(self):
    self.adjacency = {}

  def AddVertice(self, v1, v2, w):
    if self.adjacency.get(v1) == None:
      self.adjacency[v1] = []
    self.adjacency[v1].append((v2, w))

  def IsAdjacent(self, v1, v2):
    if v2 in map(lambda x: x[0], self.adjacency[v1]):
      return 'T'
    else:
      return 'F'

  def Neighbors(self, v):
    adjacency = []
    for tmp in self.adjacency[v]:
      adjacency.append(str(tmp[0]))
    return adjacency


if __name__ == "__main__":
  g = Graph()
  with open('input.txt', 'r') as inFile:
    lines = inFile.readlines()
  with open('output.txt', 'w') as outFile:
    i = 0

    while i < len(lines):
      words = lines[i].split()
      op = words[0]

      if op == CONSTRUCT:
        if len(words) != 3:
          raise Exception("CONSTRUCT: invalid input")
        
        n, m = int(words[1]), int(words[2])
        for _ in range(m):
          i += 1
          v1, v2, w = map(int, lines[i].split())
          g.AddVertice(v1, v2, w)

      elif op == IS_ADJACENT:
        if len(words) != 3:
          raise Exception("IS_ADJACENT: invalid input")
        u, v = int(words[1]), int(words[2])
        outFile.write(f'{u} {v} {g.IsAdjacent(u, v)}\n')

      elif op == GET_NEIGHBORS:
        if len(words) != 2:
          raise Exception("GET_NEIGHBORS: invalid input")
        u = int(words[1])
        outFile.write(" ".join(g.Neighbors(u)))
        outFile.write('\n')

      else:
        raise Exception("Undefined operator")
      i += 1

File: Graph/GraphPractice.cpp
// Practice 10. Graph Representation
#include <algorithm>
#include <cassert>
#include <climits>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <queue> // BFS
#include <sstream>
#include <stack> // DFS
#include <tuple>
#include <vector>
#include <map>
using namespace std;
const char CONSTRUCT = 'C';
const char IS_ADJACENT = 'I';
const char GET_NEIGHBORS = 'N';
const char BFS = 'B';
const char DFS = 'D';
const char REACHABLITY = 'R';
const char TOPOLOGICAL_SORT = 'T';
const char SHORTEST_PATH = 'S';

class Graph {
public:
int nodes;
map<int, vector<pair<int, int> > > adjacency;
// TODO. Define a constructor, a destructor, and proper methods

void AddVertice(int, int, int);
string IsAdjacent(int, int);
string Neighbors(int);
};

void Graph::AddVertice(int v1, int v2, int weight) {
  adjacency[v1].push_back(make_pair(v2, weight));
}

string Graph::IsAdjacent(int v1, int v2) {
  vector<pair<int, int> > from = adjacency[v1];
  for (vector<pair<int, int> >::iterator iter=from.begin(); iter!=from.end(); iter++) {
    if (iter->first == v2)
      return "T";
  }
  return "F";
}

string Graph::Neighbors(int v) {
  string neighbors = "";
  for (int i=0; i<adjacency[v].size(); i++)
    neighbors += to_string(adjacency[v][i].first) + " ";
  return neighbors;
}
 
int main() {
  Graph g;
  ifstream inFile("input.txt");
  ofstream outFile("output.txt");
  string line;

  while (getline(inFile, line))
  {
    char op = line[0];
    istringstream iss(line.substr(1));
    int u, v, wgt, nVertices, nEdges, cnt;
    //vector<tuple<int, int, int>> data;
    vector<int> p, nbrs;
    switch(op)
    {
      case CONSTRUCT:
        if (!(iss >> nVertices >> nEdges)) {
          cerr<<"CONSTRUCT: invalid input"<<endl;
          exit(1);
        }
        // TODO. Construct a graph
        for (int i=0; i<nEdges; i++) {
          getline(inFile, line);
          istringstream is_(line);
          is_ >> u >> v >> wgt;
          g.AddVertice(u, v, wgt);
        }
        break;

      case IS_ADJACENT:
        if (!(iss >> u >> v)) {
          cerr<<"isAdjacent: invalid input"<<endl;
          exit(1);
        }
        // TODO. Check if edge (u, v) exists in the graph
        outFile << u << " " << v << " " << g.IsAdjacent(u, v) << endl;
        break;

      case GET_NEIGHBORS:
        if (!(iss >> u)) {
          cerr<<"getNeighbors: invalid input"<<endl;
          exit(1);
        }
        // TODO. Get all the neighbors of u
        outFile << g.Neighbors(u) << endl;
        break;
        
      default:
        cerr<<"Undefined operator"<<endl;
        exit(1);
    }
  }
  outFile.close();
  inFile.close();
}
